contracts/proxy-lib/src/lib.rs

```diff
@@ -0,0 +1,194 @@
+use near_sdk::{env, log, near, AccountId, Gas, PanicOnDefault, Promise, PromiseError};
+
+pub mod ext_config;
+pub use crate::ext_config::config_contract;
+
+pub type ProposalId = u32;
+
+#[derive(PartialEq, Debug)]
+#[near(serializers = [json, borsh])]
+pub struct ProposalWithApprovals {
+    pub proposal_id: ProposalId,
+    pub num_approvals: usize,
+}
+
+#[near(contract_state)]
+#[derive(PanicOnDefault)]
+pub struct ProxyContract {
+    pub context_id: Repr<ContextId>,
+    pub context_config_account_id: AccountId,
+    pub num_approvals: u32,
+    pub proposal_nonce: ProposalId,
+    pub proposals: IterableMap<ProposalId, ProposalWithSigner>,
+    pub approvals: IterableMap<ProposalId, HashSet<Repr<SignerId>>>,
+    pub num_proposals_pk: IterableMap<Repr<SignerId>, u32>,
+    pub active_proposals_limit: u32,
+}
+
+#[derive(Clone, Debug)]
+#[near(serializers = [borsh])]
+pub struct FunctionCallPermission {
+    allowance: Option<U128>,
+    receiver_id: AccountId,
+    method_names: Vec<String>,
+}
+
+// An internal request wrapped with the signer_pk and added timestamp to determine num_requests_pk and prevent against malicious key holder gas attacks
+#[derive(Clone, PartialEq, Debug)]
+#[near(serializers = [json, borsh])]
+pub struct ProposalWithSigner {
+    pub proposal: Proposal,
+    pub signer_id: Repr<SignerId>,
+}
+
+// An internal request wrapped with the signer_pk and added timestamp to determine num_requests_pk and prevent against malicious key holder gas attacks
+#[derive(Clone, PartialEq)]
+#[near(serializers = [json, borsh])]
+pub struct ConfirmationRequestWithSigner {
+    pub proposal_id: ProposalId,
+    pub signer_id: Repr<SignerId>,
+    pub added_timestamp: u64,
+}
+
+/// Lowest level action that can be performed by the multisig contract.
+#[derive(Clone, PartialEq, Debug)]
+#[near(serializers = [json, borsh])]
+pub enum ProposalAction {
+    FunctionCall {
+        method_name: String,
+        args: Base64VecU8,
+        deposit: U128,
+        gas: U64,
+    },
+}
+
+// The request the user makes specifying the receiving account and actions they want to execute (1 tx)
+#[derive(Clone, PartialEq, Debug)]
+#[near(serializers = [json, borsh])]
+pub struct Proposal {
+    pub receiver_id: AccountId,
+    pub actions: Vec<ProposalAction>,
+}
+
+#[near]
+impl ProxyContract {
+    #[init]
+    pub fn init(context_id: Repr<ContextId>, context_config_account_id: AccountId) -> Self {
+        assert!(!env::state_exists(), "Already initialized");
+        Self {
+            context_id,
+            context_config_account_id,
+            proposal_nonce: 0,
+            proposals: IterableMap::new(b"r".to_vec()),
+            approvals: IterableMap::new(b"c".to_vec()),
+            num_proposals_pk: IterableMap::new(b"k".to_vec()),
+            num_approvals: 2,
+            active_proposals_limit: 10,
+        }
+    }
+
+    pub fn create_proposal(&mut self, proposal: Signed<ProposalWithSigner>) -> ProposalId {
```

Comment:
do we want anybody to be able submit a proposal even if they're not a context member?

Reply:
+1 it must panic and keep the map empty. Will work on it